shim/src/glue.c

```diff
@@ -1,23 +1,38 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT license.
-
+#define _GNU_SOURCE
 #include <demi/libos.h>
 #include <demi/sga.h>
 #include <demi/wait.h>
 
-static __thread int __demi_reent_guard = 0;
+#include <string.h>
+#include <unistd.h>
+#include <assert.h>
+#include <sys/types.h>
+
+#include "utils.h"
+#include "log.h"
+
+static struct hashset *__demi_reent_guards;
 
-#define DEMI_CALL(type, fn, ...)    \
-    {                               \
-        __demi_reent_guard = 1;     \
-        type ret = fn(__VA_ARGS__); \
-        __demi_reent_guard = 0;     \
-        return (ret);               \
+#define DEMI_CALL(type, fn, ...)                        \
+    {                                                   \
+        pid_t tid = gettid();                           \
+        hashset_insert(__demi_reent_guards, tid);       \
+        type ret = fn(__VA_ARGS__);                     \
+        hashset_remove(__demi_reent_guards, tid);       \
+        return (ret);                                   \
     }
 
 int is_reentrant_demi_call()
 {
-    return __demi_reent_guard;
+    pid_t tid = gettid();
+    return hashset_contains(__demi_reent_guards, tid);
+}
+
+void init_reent_guards() {
+    __demi_reent_guards = hashset_create(10);
```

Comment:
Could you create a constant for this magic value? Also, could you add some rationale behind this value? For instance, why have you cherry-picked `10`?

Reply:
The rationale was mostly that it was the same value used for the other hashsets and hash tables (1024 elements), but I believe we may be able to use a smaller number, since I imagine the number of threads will be smaller than the number of file descriptors.